src/form.rs

```diff
@@ -62,9 +65,23 @@ impl Form for MetaForm {
 /// This resolves some lifetime issues with self-referential structs (such as
 /// the registry itself) but can no longer be used to resolve to the original
 /// underlying data.
+///
+/// `type String` is owned in order to enable decoding
 #[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Serialize, Debug)]
 pub enum CompactForm {}
 
 impl Form for CompactForm {
 	type TypeId = UntrackedSymbol<TypeId>;
+	type String = &'static str;
+}
+
+/// Owned form for decoding
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Serialize, Debug)]
+pub enum OwnedForm {}
+
+impl Form for OwnedForm {
+	/// For correctness should be [`NonZeroU32`](`core::num::NonZeroU32`), but the current
+	/// (`1.3.x`) release of `parity-scale-codec` does not include the `NoneZero*` Decode impls.
+	type TypeId = u32;
```

Comment:
We should not use raw types for indices but instead something type safe using a new-type wrapper. However, this doesn't matter too much if we remove `OwnedForm` again due to my proposal above.